clippy_lints/src/match_vec_item.rs

```diff
@@ -0,0 +1,111 @@
+use crate::utils::{is_wild, span_lint_and_help};
+use if_chain::if_chain;
+use rustc_hir::{Arm, Expr, ExprKind, MatchSource};
+use rustc_lint::{LateContext, LateLintPass, LintContext};
+use rustc_middle::lint::in_external_macro;
+use rustc_middle::ty::{self, AdtDef};
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+
+declare_clippy_lint! {
+    /// **What it does:** Checks for `match vec[idx]` or `match vec[n..m]`.
+    ///
+    /// **Why is this bad?** This can panic at runtime.
+    ///
+    /// **Known problems:** None.
+    ///
+    /// **Example:**
+    /// ```rust, no_run
+    /// let arr = vec![0, 1, 2, 3];
+    /// let idx = 1;
+    ///
+    /// // Bad
+    /// match arr[idx] {
+    ///     0 => println!("{}", 0),
+    ///     1 => println!("{}", 3),
+    ///     _ => {},
+    /// }
+    /// ```
+    /// Use instead:
+    /// ```rust, no_run
+    /// let arr = vec![0, 1, 2, 3];
+    /// let idx = 1;
+    ///
+    /// // Good
+    /// match arr.get(idx) {
+    ///     Some(0) => println!("{}", 0),
+    ///     Some(1) => println!("{}", 3),
+    ///     _ => {},
+    /// }
+    /// ```
+    pub MATCH_VEC_ITEM,
+    style,
+    "match vector by indexing can panic"
+}
+
+declare_lint_pass!(MatchVecItem => [MATCH_VEC_ITEM]);
+
+impl<'a, 'tcx> LateLintPass<'a, 'tcx> for MatchVecItem {
+    fn check_expr(&mut self, cx: &LateContext<'a, 'tcx>, expr: &'tcx Expr<'tcx>) {
+        if_chain! {
+            if !in_external_macro(cx.sess(), expr.span);
+            if let ExprKind::Match(ref ex, ref arms, MatchSource::Normal) = expr.kind;
+            if contains_wild_arm(arms);
+            if is_vec_indexing(cx, ex);
+
+            then {
+                span_lint_and_help(
+                    cx,
+                    MATCH_VEC_ITEM,
+                    expr.span,
+                    "indexing vector may panic",
+                    None,
+                    "consider using `get(..)` instead.",
```

Comment:
This could easily be converted to an auto applicable suggestion. Just return the index expression from the `is_vec_indexing` function as `Option<&Expr>`  and then use `utils::snippet_with_applicability` to  build the suggestion.